space-wizards/space-station-14
# Add undergarments & "Censor Nudity" toggle to options

## About the PR


This PR adds the ability to equip undergarments to characters, as well as a Censor Nudity Toggle. This toggle is available as a client option in the Accessibility tab of the Options menu, and as a server variable.  

There are 4 top and 3 bottom undergarments added:
> Tanktop (default with toggle)
> Binder
> Classic Bra
> Sports Bra

> Boxers (default with toggle)
> Briefs
> Satin

These undergarments can be found as markings in the character creator. 

The Censor Nudity Toggle is available under the Accessibility tab in Options. If enabled, characters *who have not selected a top/bottom undergarment* will be outfitted with a default selection (Tanktop + Boxers). This default selection is overridden if another undergarment has been selected to that slot. Note that this option is clientside-only, and will not be replicated on other clients.

Servers have the ability to override the client selection via the CCvar `server_nudity_toggle`, which when set to true makes all clients behave as if they have the Censor Nudity Toggle option selected. If the CCvar is set to false, the client selection will take priority.

This PR is inspired by #31221. 

## Why / Balance


Nudity, even if it is just a single pixel or two, may cause discomfort for some players. This can manifest as discomfort at having their own character bared nude for other players, or other characters being nude around them. A nudity toggle allows for said players to engage with the game without impacting the game for other players. 

Streamers may also have issues with nudity due to streaming platform Terms of Services. 

Since nudity is not supposed to be a focus in the game (at least upstream) the selection of undergarments have been made intentionally very limited and lacks customization options. The selection should be enough for a player to select undergarments that is suitable for their character's presentation without putting undue focus on using it as a means of expression. 

Reptilians and Vox have custom undergarments due to differences in sprites (Vox are self-explanatory, Reptilians have a higher-up crotch area).

**Note that the specific sprites are not necessarily final representations, and are open to be resprited.**

## Technical details


`LayerColoringType` has been made nullable, as there was no option for a type that has a single color without a slider that isn't dependent on another property (unlike how eye/skincolor is). If it's null, it just uses the default color and won't allow it to be changed.

There's some basic options code and cvars, what's important is what happens in `HumanoidAppearanceSystem.cs`:

In `UpdateSprite()`, it checks whether nudity censoring is toggled either client- or serverside; if there already exists a marking that matches `UndergarmentTop`/`UndergarmentBottom` all is well and not much more happens, but if one is missing `AddUndergarments()` is ran and applies a marking set via `HumanoidAppearanceComponent`. These markings are also added to `ClientOldMarkings` so that they can be removed should the toggle turn off/a marking be selected in the character menu. 

## Media


Tanktop + Boxers; the default, which displays when Censor Nudity Toggle is enabled:
![image](https://github.com/user-attachments/assets/c24e52a2-d759-465d-8613-c3929ea8eb45)![image](https://github.com/user-attachments/assets/58e3750c-6822-48a0-a7be-4cbd0615edf6)![image](https://github.com/user-attachments/assets/1599da0b-82bf-4d0c-a1d5-abade6f25a14)

Classic Bra + Briefs:
![image](https://github.com/user-attachments/assets/533db4f7-f406-47a2-bccc-315437cfeb26)

Sports Bra + Satin:
![image](https://github.com/user-attachments/assets/e4b3bfe0-a669-4892-a2b3-f9b5df363a71)

Binder + Boxers:
![image](https://github.com/user-attachments/assets/6065b494-53c2-4612-b857-d095a0b61cb4)

Example lizard clothing, Tanktop + Boxers:
![image](https://github.com/user-attachments/assets/75552ff7-a0eb-4277-a4cd-34bea7313f33)

Example vox clothing, Tanktop + Boxers:
![image](https://github.com/user-attachments/assets/23d2c392-ab2c-4f18-92d7-9f7d85af1539)


![image](https://github.com/user-attachments/assets/74e980d2-5d0e-4fe5-84aa-af2583fe15b6)
![image](https://github.com/user-attachments/assets/8feb506f-5384-4dd6-8957-2de69e088c99)

## How to test

1. Boot up Dev map
2. Go into Options menu, Accessibility tab, toggle Censor Nudity Toggle on.
3. Observe how Urists in the Dev map have clothing.
4. Use the Command `golobby` to go to the lobby.
5. Enter character creation, select Hide Clothes and go to the Markings tab
6. Select a combination of undergarments and notice how they override the default clothing.
7. Save, go back to Options and toggle it off. Notice how the selected undergarments remain. 
8. Go back into the Dev map, see that Urists are naked.
9. Type the Command `> cvar accessibility.server_censor_nudity True`.
10. Observe how Urists in the Dev map have clothing again.

## Requirements



## Breaking changes


**Changelog**


:cl:
- add: Top and bottom undergarments can now be found in the Markings tab of the character creator.
- add: The Accessibility tab of Options now has a Censor Nudity Toggle, which adds undergarments to all characters on your client. Undergarments selected in the character creator take priority.
